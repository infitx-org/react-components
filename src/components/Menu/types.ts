export interface MenuSectionProps {
  label: string;
  hidden?: boolean;
  icon?: React.ReactElement<React.SVGProps<SVGSVGElement>>;
  fill?: string;
  size?: number;
  disabled?: boolean;
  children: MenuItemElement | MenuItemElement[];
}

interface BaseMenuItemProps {
  label: string;
  active?: boolean;
  disabled?: boolean;
  hidden?: boolean;
  back?: boolean;
  icon?: React.ReactElement<React.SVGProps<SVGSVGElement>>;
  fill?: string;
  size?: number;
  // prop partial is not used directly by the MenuItem but it is used by the Menu component
  partial?: boolean;
  children?: MenuItemElement | MenuElement[];
}

interface BaseMenuItemPropsWithTo extends BaseMenuItemProps {
  to: string;
  path?: string;
}
interface BaseMenuItemPropsWithPath extends BaseMenuItemProps {
  to?: string;
  path: string;
}

export type MenuItemProps = BaseMenuItemPropsWithTo | BaseMenuItemPropsWithPath;

export type MenuItemElement = React.ReactElement<MenuItemProps>;
export type MenuSectionElement = React.ReactElement<MenuSectionProps>;
export type MenuElement = MenuItemElement | MenuSectionElement;
