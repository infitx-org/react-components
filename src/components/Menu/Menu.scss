@use "../../scss/base/colors.scss";

.rc-menu {
  overflow: hidden;
  background: #f2f1f1;
}
