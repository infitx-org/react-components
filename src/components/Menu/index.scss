@use "./Menu.scss";
@use "./components/MenuItem.scss";
@use "./components/MenuSection.scss";
