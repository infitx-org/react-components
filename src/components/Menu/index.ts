export { default } from "./Menu";
export { MenuProps } from "./Menu";
