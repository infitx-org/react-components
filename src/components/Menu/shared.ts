
import React from "react";
import MenuItem from './components/MenuItem';
import MenuSection from './components/MenuSection';

export interface MenuSectionProps {
  label: string;
  hidden?: boolean;
  icon?: string;
  fill?: string;
  size?: number;
  disabled?: boolean;
  children: MenuItemElement[];
}

export interface MenuItemProps {
  label: string;
  active: boolean;
  path?: string;
  to?: string;
  disabled?: boolean;
  hidden?: boolean;
  back?: boolean;
  icon?: string;
  fill?: string;
  size?: number;
  // prop asRoot is not used directly by the MenuItem but it is used by the Menu component
  // eslint-disable-next-line react/no-unused-prop-types
  asRoot?: boolean;
  // prop partial is not used directly by the MenuItem but it is used by the Menu component
  // eslint-disable-next-line react/no-unused-prop-types
  partial?: boolean;
  // eslint-disable-next-line react/no-unused-prop-types
  children?: MenuElement[];
}

export type MenuItemElement = React.ReactElement<MenuItemProps>;
export type MenuSectionElement = React.ReactElement<MenuSectionProps>;
export type MenuElement = MenuItemElement | MenuSectionElement;

export interface MenuContextValue {
  pathname: string;
  onClick: (path: string) => void;
}

export const MenuContext = React.createContext<MenuContextValue>({
  pathname: "",
  onClick: () => undefined,
});

export function getPathMatches(
  pathname: string,
  path: string | undefined,
  partial: boolean | undefined
): boolean {
  if (!path) {
    return false;
  }
  if (partial && pathname.startsWith(path)) {
    return true;
  }
  if (path === pathname) {
    return true;
  }

  const pathChunks = path.split("/");
  const pathNameChunks = pathname.split("/");

  if (pathChunks.length !== pathNameChunks.length) {
    return false;
  }

  return pathNameChunks.every((chunk, index) => {
    const pathChunk = pathChunks[index];
    const isExact = pathChunk === chunk;
    const isWild = pathChunk !== undefined && pathChunk.startsWith(":");
    return isExact || isWild;
  });
}

export function isMenuItem(child: React.ReactNode): child is MenuItemElement {
  return (child as React.ReactElement).type === MenuItem;
}

export function isMenuSection(child: React.ReactNode): child is MenuSectionElement {
  return (child as React.ReactElement).type === MenuSection;
}