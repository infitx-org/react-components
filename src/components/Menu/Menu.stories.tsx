import React, { PureComponent } from "react";
import Row from "components/Layout/Row";
import TextField from "components/TextField";
import Checkbox from "components/Checkbox";
import Button from "components/Button";
import Menu from "./Menu";

export default {
  title: "Components/Menu",
  component: Menu,
  subcomponents: {
    "Menu.Item": Menu.Item,
    "Menu.Section": Menu.Section,
  }, // ðŸ‘ˆ Adds the ListItem component as a subcomponent
};

const icon = (
  <svg width="40" height="40" viewBox="0 0 40 40">
    <circle cx="20" cy="20" r="20" />
  </svg>
);

export const Default = () => {
  const [pathname, setPathname] = React.useState("/");
  return (
    <Menu path="/" pathname={pathname} onChange={setPathname}>
      <Menu.Section label="Items">
        <Menu.Item label="One" path="/one" />
        <Menu.Item label="Two" path="/two" />
      </Menu.Section>
      <Menu.Section label="Settings">
        <Menu.Item label="Preferences" path="/preferences" />
        <Menu.Item label="Account" path="/account" />
      </Menu.Section>
    </Menu>
  );
};

export const MenuItemWithIcon = () => {
  return <Menu.Item label="Two" path="/two" icon={icon} />;
};

export const WithSection = () => {
  const [pathname, setPathname] = React.useState("/");
  return (
    <Menu path="/" pathname={pathname} onChange={setPathname}>
      <Menu.Section label="Items">
        <Menu.Item label="One" path="/one" />
        <Menu.Item label="Two" path="/two" />
      </Menu.Section>
    </Menu>
  );
};

export const Nested = () => {
  const [pathname, setPathname] = React.useState("/");
  return (
    <Menu path="/" pathname={pathname} onChange={setPathname}>
      <Menu.Section label="Items">
        <Menu.Item label="One" path="/one" />
        <Menu.Item label="Two" path="/two">
          <Menu.Item label="Go Back" path="/" back />
          <Menu.Item label="Nested" path="/two/test" />
        </Menu.Item>
      </Menu.Section>
    </Menu>
  );
};

export const PartialPath = () => {
  const [pathname, setPathname] = React.useState("/user/account");
  return (
    <Menu path="/" pathname={pathname} onChange={setPathname}>
      <Menu.Item label="One" path="/one" />
      <Menu.Item label="Two" path="/two" />
      <Menu.Item label="User" path="/user" partial />
    </Menu>
  );
};

export const ParametrizedPath = () => {
  const [pathname, setPathname] = React.useState("/users/1/comments");
  return (
    <Menu path="/" pathname={pathname} onChange={setPathname}>
      <Menu.Item label="One" path="/one" />
      <Menu.Item label="Two" path="/two" />
      <Menu.Item label="User Comments" path="/users/:id/comments" />
    </Menu>
  );
};
