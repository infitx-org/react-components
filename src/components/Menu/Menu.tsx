import React from "react";
import MenuItem from "./components/MenuItem";
import MenuSection from "./components/MenuSection";
import { isActivePath, MenuContext } from "./shared";
import { MenuElement, MenuItemElement, MenuSectionElement } from "./types";
import "./Menu.scss";

export function isMenuItem(child: React.ReactNode): child is MenuItemElement {
  return (child as React.ReactElement).type === MenuItem;
}

export function isMenuSection(
  child: React.ReactNode
): child is MenuSectionElement {
  return (child as React.ReactElement).type === MenuSection;
}

function stripMenuSections(
  items: MenuItemElement | MenuElement[]
): MenuElement[] {
  const inArray = React.Children.toArray(items) as MenuElement[];
  // It flattens nested MenuSection, MenuItem components into a one-level array
  return inArray.reduce(
    (prevItems, currentItem) => [
      ...prevItems,
      ...(isMenuSection(currentItem)
        ? stripMenuSections(currentItem.props.children)
        : [currentItem]),
    ],
    [] as MenuElement[]
  );
}

function findVisibleElements(
  menuElements: MenuItemElement | MenuElement[],
  pathname: string | undefined
): MenuItemElement | MenuElement[] | undefined {
  let activeNode: MenuItemElement | MenuElement[] | undefined;

  // strip Menu.Section components in order to
  // flat children when detecting active menu
  stripMenuSections(menuElements)
    .filter(isMenuItem)
    .some((menuItem) => {
      // find the first matching menu item and return the parent or the item itself
      // depending if needs to be treated like a root
      const { partial, path, active } = menuItem.props;
      if (active || isActivePath(pathname, path, partial)) {
        // isRoot prop is meant to be used when menu has child elements
        // and we do not want to render the parent menuItem but the child menuItems
        const isRoot = menuItem.props.children !== undefined;
        activeNode = isRoot ? menuItem.props.children : menuElements;
      } else if (menuItem.props.children !== undefined) {
        activeNode = findVisibleElements(menuItem.props.children, pathname);
      }

      return !!activeNode;
    });

  return activeNode;
}

export interface MenuProps {
  path: string;
  pathname?: string;
  children: MenuItemElement | MenuElement[];
  onChange: (p: string) => void;
}

function Menu({ path, pathname, onChange, children }: MenuProps) {
  // Default to Menu
  let menuElements: MenuItemElement | MenuElement[] | undefined = children;
  if (pathname !== path) {
    menuElements = findVisibleElements(children, pathname) || children;
  }
  const menuComponents = React.Children.toArray(menuElements).filter(
    (element) => isMenuItem(element) || isMenuSection(element)
  ) as MenuElement[];

  return (
    <MenuContext.Provider value={{ pathname, onClick: onChange }}>
      <div className="rc-menu">{menuComponents}</div>
    </MenuContext.Provider>
  );
}

Menu.Item = MenuItem;
Menu.Section = MenuSection;

export default Menu;
