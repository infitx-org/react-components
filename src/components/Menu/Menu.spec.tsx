import { render, screen } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";
import userEvent from "@testing-library/user-event";

import Menu from "./Menu";

const icon = (
  <svg width="40" height="40" viewBox="0 0 40 40">
    <circle cx="20" cy="20" r="20" />
  </svg>
);

const onChangeMockEvent = jest.fn();

it("renders only the menu", () => {
  const { container } = render(
    <Menu path="/" pathname="/" onChange={onChangeMockEvent} />
  );
  expect(container.querySelectorAll(".rc-menu")).toHaveLength(1);
});

it("renders the menu wrapper also if pathname does not match any route", () => {
  const { container } = render(
    <Menu path="/" pathname="/test" onChange={onChangeMockEvent} />
  );
  expect(container.querySelectorAll(".rc-menu")).toHaveLength(1);
});

it("renders the menu items if parent root matches", () => {
  const { container } = render(
    <Menu path="/" pathname="/" onChange={onChangeMockEvent}>
      <Menu.Item path="/foo" label="foo" />
      <Menu.Item path="/bar" label="bar" />
    </Menu>
  );
  expect(container.querySelectorAll(".rc-menu-item")).toHaveLength(2);
});

it("renders the menu items if child route matches", () => {
  const { container } = render(
    <Menu path="/" pathname="/foo" onChange={onChangeMockEvent}>
      <Menu.Item path="/foo" label="foo" />
      <Menu.Item path="/bar" label="bar" />
    </Menu>
  );
  expect(container.querySelectorAll(".rc-menu-item")).toHaveLength(2);
});

it("renders the menu section with icons", () => {
  const { container } = render(
    <Menu path="/" pathname="/foo" onChange={onChangeMockEvent}>
      <Menu.Section label="Menu Section" icon={icon}>
        <Menu.Item path="/foo" label="foo" />
        <Menu.Item path="/bar" label="bar" />
      </Menu.Section>
    </Menu>
  );
  expect(container.querySelectorAll(".rc-menu-section__icon")).toHaveLength(1);
});

it("renders the menu items with icons", () => {
  const { container } = render(
    <Menu path="/" pathname="/foo" onChange={onChangeMockEvent}>
      <Menu.Section label="Menu Section">
        <Menu.Item path="/foo" label="foo" icon={icon} />
        <Menu.Item path="/bar" label="bar" icon={icon} />
      </Menu.Section>
    </Menu>
  );
  expect(container.querySelectorAll(".rc-menu-item__item-icon")).toHaveLength(
    2
  );
});

// it('does not render the menu items if no route matches', () => {
//   const { container } = render(
//     <Menu path="/" pathname="/non-existing" onChange={onChangeMockEvent}>
//       <Menu.Item path="/foo" label="foo" />
//       <Menu.Item path="/bar" label="bar" />
//     </Menu>,
//   );
//   expect(container.querySelectorAll('.rc-menu-item')).toHaveLength(0);
// });

it("renders the nested child when asRoot prop is set and route matches parent", () => {
  const { container } = render(
    <Menu path="/" pathname="/foo" onChange={onChangeMockEvent}>
      <Menu.Item path="/foo" label="foo">
        <Menu.Item path="/foo/nested" label="nested" />
      </Menu.Item>
      <Menu.Item path="/bar" label="bar" />
    </Menu>
  );
  const menuItem = container.querySelector(".rc-menu-item");
  expect(menuItem).toBeTruthy();
  expect(menuItem?.textContent).toBe("nested");
  expect(container.querySelectorAll(".rc-menu-item")).toHaveLength(1);
});

it("renders the nested child if matching if asRoot prop is not set ", () => {
  const { container } = render(
    <Menu path="/" pathname="/1st/2nd/3rd/4th" onChange={onChangeMockEvent}>
      <Menu.Item path="/1st" label="1st Level">
        <Menu.Item path="/1st/2nd" label="2st level">
          <Menu.Item path="/1st/2nd/3rd" label="3nd level">
            <Menu.Item path="/1st/2nd/3rd/4th" label="4th level" />
          </Menu.Item>
        </Menu.Item>
      </Menu.Item>
    </Menu>
  );

  const menuItem = container.querySelector(".rc-menu-item");
  expect(menuItem).toBeTruthy();
  expect(menuItem?.textContent).toBe("4th level");
});

it("renders the menu items when partial match is allowed", () => {
  const { container } = render(
    <Menu path="/" pathname="/foo/partial/match" onChange={onChangeMockEvent}>
      <Menu.Item path="/foo" label="foo" partial />
      <Menu.Item path="/bar" label="bar" />
    </Menu>
  );
  const menuItem = container.querySelectorAll(".rc-menu-item");
  const activeMenuItem = container.querySelector(".rc-menu-item--active");
  expect(menuItem).toHaveLength(2);
  expect(activeMenuItem?.textContent).toBe("foo");
});

it("renders the menu items when router param is used", () => {
  const { container } = render(
    <Menu
      path="/"
      pathname="/users/:user-id/settings"
      onChange={onChangeMockEvent}
    >
      <Menu.Item path="/books" label="Books" />
      <Menu.Item path="/users/:user-id" label="User">
        <Menu.Item path="/users/:user-id/account" label="account" />
        <Menu.Item path="/users/:user-id/settings" label="settings" />
      </Menu.Item>
    </Menu>
  );
  const menuItem = container.querySelectorAll(".rc-menu-item");
  const activeMenuItem = container.querySelector(".rc-menu-item--active");
  expect(menuItem).toHaveLength(2);
  expect(activeMenuItem?.textContent).toBe("settings");
});

it("renders the menu section if parent root matches", () => {
  const { container } = render(
    <Menu path="/" pathname="/" onChange={onChangeMockEvent}>
      <Menu.Section label="Menu Section">
        <Menu.Item path="/foo" label="foo" />
        <Menu.Item path="/bar" label="bar" />
      </Menu.Section>
    </Menu>
  );
  const menuSection = container.querySelector(".rc-menu-section");
  expect(menuSection).toBeTruthy();
  const menuSectionLabel = menuSection?.querySelector(
    ".rc-menu-section__label"
  );
  expect(menuSectionLabel?.textContent).toBe("Menu Section");
});

it("renders the Menu when no pathname is set", () => {
  const { container } = render(
    <Menu onChange={onChangeMockEvent} path="?">
      <Menu.Item label="foo" path="/foo" />
      <Menu.Item label="bar" path="/bar" />
    </Menu>
  );
  const menuItems = container.querySelectorAll(".rc-menu-item");
  expect(menuItems).toHaveLength(2);
});

// it('renders the manually set active prop', () => {
//   const { container } = render(
//     <Menu onChange={onChangeMockEvent}>
//       <Menu.Item label="foo" active />
//       <Menu.Item label="bar" />
//     </Menu>,
//   );
//   const activeMenuItem = container.querySelectorAll('.rc-menu-item--active');
//   expect(activeMenuItem).toHaveLength(1);
//   expect(activeMenuItem.text()).toBe('foo');
// });

// it('renders the disabled prop on menu item', () => {
//   const { container } = render(
//     <Menu onChange={onChangeMockEvent}>
//       <Menu.Item label="foo" disabled />
//       <Menu.Item label="bar" />
//     </Menu>,
//   );
//   const disabledMenuItem = container.querySelectorAll('.rc-menu-item--disabled');
//   expect(disabledMenuItem).toHaveLength(1);
//   expect(disabledMenuItem.text()).toBe('foo');
// });

// it('does not render a hidden menu item', () => {
//   const { container } = render(
//     <Menu onChange={onChangeMockEvent}>
//       <Menu.Item label="foo" hidden />
//       <Menu.Item label="bar" />
//     </Menu>,
//   );
//   const menuItems = container.querySelectorAll('.rc-menu-item');
//   expect(menuItems).toHaveLength(1);
// });

// it('does not render a hidden menu section', () => {
//   const { container } = render(
//     <Menu onChange={onChangeMockEvent}>
//       <Menu.Section hidden>
//         <Menu.Item label="foo" />
//         <Menu.Item label="bar" />
//       </Menu.Section>
//     </Menu>,
//   );
//   const menuSection = container.querySelectorAll('.rc-men-section');
//   expect(menuSection).toHaveLength(0);
// });

// it('trigger onChange when clicking a menu item', () => {
//   const mockEvent = jest.fn();
//   const { container } = render(
//     <Menu path="/" pathname="/" onChange={mockEvent}>
//       <Menu.Item path="/foo" label="foo" />
//       <Menu.Item path="/bar" label="bar" />
//     </Menu>,
//   );
//   const menuItems = container.querySelectorAll('.rc-menu-item');
//   menuItems.at(0).simulate('click');
//   expect(mockEvent).toHaveBeenCalled();
// });

// it('trigger onChange with correct value when clicking a menu item', () => {
//   const mockEvent = jest.fn();
//   const { container } = render(
//     <Menu path="/" pathname="/" onChange={mockEvent}>
//       <Menu.Item path="/foo" label="foo" />
//       <Menu.Item path="/bar" label="bar" />
//     </Menu>,
//   );
//   const menuItems = container.querySelectorAll('.rc-menu-item');
//   menuItems.at(1).simulate('click');
//   expect(mockEvent).toHaveBeenCalledWith('/bar');
// });

// // Snapshot
// it('renders the menu correctly when multiple props are set', () => {
//   const { container } = render(
//     <Menu path="/" pathname="/foo" onChange={onChangeMockEvent}>
//       <Menu.Item path="/foo" label="foo" asRoot>
//         <Menu.Item path="/foo/nested" label="nested" />
//       </Menu.Item>
//       <Menu.Item path="/bar" label="bar" />
//     </Menu>,
//   );
//   expect(shallowToJson(wrapper)).toMatchSnapshot();
// });
