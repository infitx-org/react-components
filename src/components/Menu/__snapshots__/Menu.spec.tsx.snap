// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the menu correctly when multiple props are set 1`] = `
<div>
  <div
    class="rc-menu"
  >
    <div
      class="rc-menu-item"
      role="presentation"
    >
      nested
    </div>
  </div>
</div>
`;
